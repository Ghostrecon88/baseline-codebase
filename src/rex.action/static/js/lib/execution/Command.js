/**
 * @copyright 2015, Prometheus Research, LLC
 */

import React            from 'react';
import invariant        from 'invariant';
import * as Entity      from '../Entity';
import notImplemented   from '../notImplemented';
import * as StringUtils from '../StringUtils';

export function command(...argumentTypes) {
  return function command__decorate(target, key, desc) {
    return {
      ...desc,
      value: new Command(desc.value, key, argumentTypes)
    };
  };
}

export function getCommand(actionElement, commandName) {
  invariant(
    React.isValidElement(actionElement),
    'Expected a valid React element'
  );
  if (commandName === onContextCommand.name) {
    return onContextCommand;
  }
  return actionElement.type.commands ?
    actionElement.type.commands[commandName] :
    null;
}


export class Command {

  constructor(execute, name, argumentTypes) {
    this._execute = execute;
    this.name = name;
    this.argumentTypes = argumentTypes;
  }

  execute(...args) {
    return this._execute(...args);
  }

}

class ArgumentType {

  @notImplemented
  parse(_actionElement, _value) {
    /* istanbul ignore next */
  }

  @notImplemented
  stringify(_actionElement, _value) {
    /* istanbul ignore next */
  }

  @notImplemented
  check(_actionElement, _value) {
    /* istanbul ignore next */
  }
}

class ValueType extends ArgumentType {

  stringify(actionElement, value) {
    if (value == null) {
      return '~';
    } else {
      return String(value);
    }
  }

  parse(actionElement, value) {
    if (value === '~') {
      return null;
    } else {
      return value;
    }
  }

}

class EntityType extends ArgumentType {

  stringify(actionElement, entity) {
    if (entity == null) {
      return '';
    }
    invariant(
      Entity.isEntity(entity),
      'Expected an entity, got: %s', entity
    );
    let type = Entity.getEntityType(entity);
    let value = type + ':' + entity.id;
    let state = Entity.getEntityState(entity);
    let stateKeys = Object.keys(state);
    if (stateKeys.length > 0) {
      value = value + '!' + stateKeys.join(',');
    }
    return value;
  }

  parse(actionElement, value) {
    if (value === '') {
      return null;
    }
    let [type, value2] = value.split(':');
    if (value2.indexOf('!') > -1) {
      let [id, stateKeys] = value2.split('!');
      let state = {};
      stateKeys.split(',').forEach(key => state[key] = true);
      return Entity.createEntity(type, id, null, state);
    } else {
      return Entity.createEntity(type, value2);
    }
  }
}

class ConfigurableEntityType extends ArgumentType {

  constructor(propName) {
    super();
    this.propName = propName;
  }

  stringify(actionElement, entity) {
    if (entity == null) {
      return '';
    }
    invariant(
      Entity.isEntity(entity),
      'Expected an entity, got: %s', entity
    );
    let value = entity.id;
    let state = Entity.getEntityState(entity);
    let stateKeys = Object.keys(state);
    if (stateKeys.length > 0) {
      value = value + '!' + stateKeys.join(',');
    }
    return value;
  }

  parse(actionElement, value) {
    if (value === '') {
      return null;
    }
    let type = actionElement.props[this.propName];
    invariant(
      type,
      'Action should have prop "%s" defined', this.propName
    );
    if (value.indexOf('!') > -1) {
      let [id, stateKeys] = value.split('!');
      let state = {};
      stateKeys.split(',').forEach(key => state[key] = true);
      return Entity.createEntity(type.type.name, id, null, state);
    } else {
      return Entity.createEntity(type.type.name, value);
    }
  }
}

export class ObjectArgument extends ArgumentType {

  stringify(actionElement, object) {
    let value = [];
    for (let k in object) {
      if (object.hasOwnProperty(k)) {
        let v = object[k];
        if (v == null) {
          continue;
        }
        if (Entity.isEntity(v)) {
          v = '~' + entityType.stringify(actionElement, v);
        }
        value.push(StringUtils.joinWithEquals([k, v]));
      }
    }
    return StringUtils.joinWithComma(value);
  }

  parse(actionElement, value) {
    let object = {};
    let kv = StringUtils.splitByComma(value);
    for (let i = 0; i < kv.length; i++) {
      let [k, v] = StringUtils.splitByEquals(kv[i]);
      if (v[0] === '~') {
        v = entityType.parse(actionElement, v.substring(1));
      }
      object[k] = v;
    }
    return object;
  }
}

let entityType = new EntityType();
let valueType = new ValueType();

export let onContextCommand = new Command(
  function(props, context, update) {
    return {...context, ...update};
  },
  'context',
  [new ObjectArgument()]
);


export let Types = {

  Value() {
    return valueType;
  },

  Entity() {
    return entityType;
  },

  ConfigurableEntity(propName = 'entity') {
    return new ConfigurableEntityType(propName);
  },

};

