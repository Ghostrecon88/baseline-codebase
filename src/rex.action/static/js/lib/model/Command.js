/**
 * @copyright 2015-present, Prometheus Research, LLC
 * @flow
 */

import type {Entity, Context} from './types';

import React from 'react';
import invariant from 'invariant';
import notImplemented from '../notImplemented';
import * as StringUtils from '../StringUtils';
import * as E from './Entity';

export function command(...argumentTypes: ArgumentType[]): Function {
  return function command__decorate(target, key, desc) {
    return {
      ...desc,
      value: new Command(desc.value, key, argumentTypes),
    };
  };
}

export function defineCommand(
  ActionComponent: Class<*>,
  spec: {
    name?: string,
    argumentTypes: ArgumentType[],
    execute: Function,
  },
): void {
  const {name = 'default', argumentTypes, execute} = spec;
  ActionComponent.commands = ActionComponent.commands || {};
  ActionComponent.commands[name] = new Command(execute, name, argumentTypes);
}

export function getCommand(
  actionElement: React$Element<*>,
  commandName: string,
): ?Command {
  invariant(React.isValidElement(actionElement), 'Expected a valid React element');
  if (commandName === onContextCommand.name) {
    return onContextCommand;
  }
  return actionElement.type.commands ? actionElement.type.commands[commandName] : null;
}

export class Command {
  _execute: Function;
  name: string;
  argumentTypes: Array<ArgumentType>;

  constructor(execute: Function, name: string, argumentTypes: Array<ArgumentType>) {
    this._execute = execute;
    this.name = name;
    this.argumentTypes = argumentTypes;
  }

  parseArguments(props: Object, args: any[]): any[] {
    return args.map((arg, idx) => this.argumentTypes[idx].parse(props, arg));
  }

  execute(props: Object, context: Object, args: any[]): Context {
    return this._execute(props, context, ...args);
  }
}

class ArgumentType {
  @notImplemented parse(_actionElement, _value) {
    /* istanbul ignore next */
  }

  @notImplemented stringify(_actionElement, _value) {
    /* istanbul ignore next */
  }

  @notImplemented check(_actionElement, _value) {
    /* istanbul ignore next */
  }
}

class ValueType extends ArgumentType {
  stringify(_props: Object, value: any) {
    if (value == null) {
      return '~';
    } else {
      return String(value);
    }
  }

  parse(_props: Object, value: string) {
    if (value === '~') {
      return null;
    } else {
      return value;
    }
  }
}

class EntityType extends ArgumentType {
  stringify(props: Object, entity: ?Entity) {
    if (entity == null) {
      return '';
    }
    invariant(E.isEntity(entity), 'Expected an entity, got: %s', entity);
    let type = E.getEntityType(entity);
    let value = type + ':' + entity.id;
    return value;
  }

  parse(props: Object, value: string) {
    if (typeof value !== 'string') {
      return value;
    }
    if (value === '') {
      return null;
    }
    let [type, value2] = value.split(':');
    return E.createEntity(type, value2);
  }
}

class ConfigurableEntityType extends ArgumentType {
  propName: string;

  constructor(propName: string) {
    super();
    this.propName = propName;
  }

  stringify(props: Object, entity: ?Entity): string {
    if (entity == null) {
      return '';
    }
    invariant(E.isEntity(entity), 'Expected an entity, got: %s', entity);
    let value = entity.id;
    return String(value);
  }

  parse(props: Object, value: string): ?Entity {
    if (typeof value !== 'string') {
      return value;
    }
    if (value === '') {
      return null;
    }
    let type = props[this.propName];
    invariant(type, 'Action should have prop "%s" defined', this.propName);
    return E.createEntity(type.type.name, value);
  }
}

export class ObjectArgument extends ArgumentType {
  stringify(props: Object, object: Object): string {
    let value = [];
    for (let k in object) {
      if (object.hasOwnProperty(k)) {
        let v = object[k];
        if (v == null) {
          continue;
        }
        if (E.isEntity(v)) {
          v = '~' + entityType.stringify(props, v);
        }
        value.push(StringUtils.joinWithEquals([k, v]));
      }
    }
    return StringUtils.joinWithComma(value);
  }

  parse(props: Object, value: string): Object {
    if (typeof value !== 'string') {
      return value;
    }
    let object = {};
    let kv = StringUtils.splitByComma(value);
    for (let i = 0; i < kv.length; i++) {
      let [k, v] = StringUtils.splitByEquals(kv[i]);
      if (v[0] === '~') {
        v = entityType.parse(props, v.substring(1));
      }
      object[k] = v;
    }
    return object;
  }
}

let entityType = new EntityType();
let valueType = new ValueType();

export let onContextCommand = new Command(
  function(props, context, update) {
    return {...context, ...update};
  },
  'context',
  [new ObjectArgument()],
);

export let Types = {
  Value() {
    return valueType;
  },

  Entity() {
    return entityType;
  },

  ConfigurableEntity(propName: string = 'entity') {
    return new ConfigurableEntityType(propName);
  },
};
