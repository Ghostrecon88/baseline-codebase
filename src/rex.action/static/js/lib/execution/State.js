/**
 * @copyright 2017-present, Prometheus Research, LLC
 * @flow
 */

import invariant from 'invariant';
import type {RecordType} from '../Typing';
import {type Command, onContextCommand} from './Command';

/**
 * Abstract instruction.
 */
export type Instruction = IInclude | IExecute | IReplace | IRepeat;

/**
 * Instruction which can contain other instructions.
 */
export type ContainerInstruction = IInclude | IRepeat;

/**
 * Execute an action (render a piece of UI to a user).
 */
export type IExecute = {
  type: 'execute',
  action: ActionSpec,
  then: Instruction[],
};

export type ActionSpec = {
  id: string,
  name?: string,
  title?: string,
  domain: Domain,
  element: React$Element<*>,
  commands: {
    [commandName: string]: Command,
  },
  contextTypes: {
    input: RecordType,
    output: RecordType,
  },
};

/**
 * Include another set of instructions.
 */
export type IInclude = {
  type: 'include',
  include: Instruction[],
  then: Instruction[],
};

/**
 * Repeat a set of instructions.
 */
export type IRepeat = {
  type: 'repeat',
  repeat: Instruction[],
  then: Instruction[],
};

/**
 * Replace the with the reference and the context update.
 */
export type IReplace = {
  type: 'replace',
  traverseBack: number,
  traverse: Array<{
    actionId: string,
    contextUpdate: Context,
  }>,
};

export type Context = {
  [key: string]: ContextValue,
};

// This is actually $key but we won't encode it in type system
type ContextKey = string;

type ContextValue = string | number | boolean;

type ContextUpdateSpec = {
  [key: string]: ContextValue | ContextKey,
};

export type Config = {
  context: Context,
  instruction: Instruction[],
};

export type Domain = {
  [entityName: string]: {
    [stateName: string]: {
      expression: Function,
    },
  },
};

export type ActionState = {
  [name: string]: string,
};

export type Position = {
  context: Context,
  state: ActionState,
  command?: {commandName: string, args: string[]},

  instruction: IExecute,
  stack: ContainerInstruction[],
  trace: Position[],
};

export type State = {
  position: ?Position,
  config: Config,
};

export function create(config: Config): State {
  return {
    position: null,
    config,
  };
}

function next(state: State): Position[] {
  if (state.position != null) {
    return nextPosition(state.position);
  } else {
    const context = state.config.context || {};
    const possibelStartPositions = [];
    collectFromTraverse(
      possibelStartPositions,
      [],
      [],
      context,
      state.config.instruction,
    );
    return possibelStartPositions;
  }
}

export function close(state: State, _actionId: string): State {
  // TODO: implement
  return state;
}

export function returnTo(state: State, actionId: string): State {
  const currentPos = state.position;
  invariant(currentPos != null, 'Invalid state');
  const nextPos = currentPos.trace.find(pos => pos.instruction.action.id === actionId);
  invariant(nextPos != null, 'Cannot return to %s: no such action in trace', actionId);
  return {...state, position: nextPos};
}

export function advanceTo(
  state: State,
  actionId: string,
  contextUpdate: ?Context,
): State {
  const nextPos = next(state)
    .map(pos => updatePositionContext(pos, contextUpdate))
    .filter(isPositionAllowed)
    .find(pos => pos.instruction.action.id === actionId);
  invariant(nextPos != null, 'Unable to advance wizard to position: %s', actionId);
  return {...state, position: nextPos};
}

export function advanceToFirst(state: State, contextUpdate?: Context): State {
  const nextPosList = next(state).map(pos => updatePositionContext(pos, contextUpdate));
  const nextPosListAllowed = nextPosList.filter(isPositionAllowed);
  invariant(nextPosListAllowed.length > 0, 'Unable to advance wizard');
  const nextPos = nextPosListAllowed[0];
  return {...state, position: nextPos};
}

export function replaceCurrentPositionWithSibling(state: State, actionId: string): State {
  const currentPos = state.position;
  invariant(currentPos != null, 'Invalid state');
  if (currentPos.trace.length === 0) {
    return state;
  }
  const prevPos = currentPos.trace[currentPos.trace.length - 1];
  const nextState = {...state, position: prevPos};
  return advanceTo(nextState, actionId);
}

export function setStateAtCurrentPosition(state: State, stateUpdate: Object): State {
  invariant(state.position != null, 'Invalid state');
  const nextPosition = {
    ...state.position,
    state: {...state.position.state, ...stateUpdate},
  };
  return {...state, position: nextPosition};
}

export function applyCommandAtCurrentPosition(
  state: State,
  commandName: string,
  args: Array<string>,
): State {
  invariant(state.position != null, 'Invalid state');
  const {instruction: {action}, context} = state.position;
  const command = getActionCommand(action, commandName);
  invariant(command != null, 'Unable to locate command: %s', commandName);
  const nextContext = command.execute(action.element.props, context, ...args);
  const nextPosition = {
    ...state.position,
    context: nextContext,
    command: {commandName, args},
  };
  return {...state, position: nextPosition};
}

export function getActionCommand(action: ActionSpec, commandName: string): ?Command {
  if (commandName === onContextCommand.name) {
    return onContextCommand;
  }
  return action.commands ? action.commands[commandName] : null;
}

/**
 * Check if the position is allowed.
 */
export function isPositionAllowed(pos: Position): boolean {
  const {contextTypes, domain} = pos.instruction.action;
  return contextTypes.input.match(pos.context, domain);
}

function updatePositionContext(pos: Position, contextUpdate: ?Context): Position {
  if (contextUpdate == null) {
    return pos;
  }
  return {...pos, context: {...pos.context, ...contextUpdate}};
}

/**
 * Produce a list of possible next positions.
 */
export function nextPosition(pos: Position): Position[] {
  const result = [];
  const nextTrace = pos.trace.concat(pos);
  if (pos.instruction.then.length > 0) {
    collectFromTraverse(result, nextTrace, pos.stack, pos.context, pos.instruction.then);
  } else if (pos.stack.length > 0) {
    collectFromStack(result, nextTrace, pos.stack, pos.context);
  }
  return result;
}

/**
 * Produce a list of possible sibling positions.
 */
export function siblingPosition(pos: Position): Position[] {
  if (pos.trace.length === 0) {
    return [];
  }
  const prev = pos.trace[pos.trace.length - 1];
  return nextPosition(prev);
}

function collectFromStack(
  to: Position[],
  trace: Position[],
  stack: ContainerInstruction[],
  context: Context,
) {
  const nextStack = stack.slice(0);
  while (nextStack.length > 0) {
    const parent = nextStack.pop();
    if (parent.type === 'include') {
      if (parent.then.length > 0) {
        collectFromTraverse(to, trace, nextStack, context, parent.then);
        break;
      }
    } else if (parent.type === 'repeat') {
      collectFromTraverse(to, trace, nextStack.concat(parent), context, parent.repeat);
      collectFromTraverse(to, trace, nextStack, context, parent.then);
      break;
    } else {
      invariant(false, 'Unable to process: %s', parent.type);
    }
  }
}

function collectFromTraverse(
  to: Position[],
  trace: Position[],
  stack: ContainerInstruction[],
  context: Context,
  instructionList: Instruction[],
) {
  for (let i = 0; i < instructionList.length; i++) {
    const instruction = instructionList[i];
    if (instruction.type === 'execute') {
      to.push({instruction, stack, trace, context, state: {}});
    } else if (instruction.type === 'include') {
      collectFromTraverse(
        to,
        trace,
        stack.concat(instruction),
        context,
        instruction.include,
      );
    } else if (instruction.type === 'repeat') {
      collectFromTraverse(
        to,
        trace,
        stack.concat(instruction),
        context,
        instruction.repeat,
      );
    } else if (instruction.type === 'replace') {
      // traverse back
      const nextTrace = trace.slice(0, trace.length - instruction.traverseBack);
      let pos = nextTrace.pop();
      invariant(pos != null, 'Unable to process replace');
      // traverse forward
      for (let i = 0; i < instruction.traverse.length; i++) {
        const {actionId, contextUpdate} = instruction.traverse[i];
        pos = nextPosition(pos).find(pos => pos.instruction.action.id === actionId);
        invariant(pos != null, 'Unable to process replace');
        // TODO: we need to refer to original context here
        pos = {...pos, context: updateContextBySpec(pos.context, context, contextUpdate)};
      }
      collectFromTraverse(to, pos.trace, pos.stack, pos.context, [pos.instruction]);
    } else {
      invariant(false, 'Unable to process: %s', instruction.type);
    }
  }
}

function updateContextBySpec(
  context: Context,
  originalContext: Context,
  updateSpec: ContextUpdateSpec,
): Context {
  const nextContext = {...context};
  for (const key in updateSpec) {
    const value = updateSpec[key];
    if (typeof value === 'string' && value.startsWith('$')) {
      nextContext[key] = originalContext[value.slice(1)];
    } else {
      nextContext[key] = value;
    }
  }
  return nextContext;
}
