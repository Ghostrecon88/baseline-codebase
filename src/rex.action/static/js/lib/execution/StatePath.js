/**
 * @copyright 2016-present, Prometheus Research, LLC
 * @flow
 */

import invariant from 'invariant';
import {splitBySlash, splitByComma, joinWithSlash, joinWithComma} from '../StringUtils';
import * as Environment from '../Environment';
import * as S from './State';

/**
 * Deserialize state from `path`.
 */
export function fromPath(path: string, config: S.Config): S.State {
  // We need a workaround as FF doesn't behave consistently.
  if (Environment.isFirefox()) {
    /* istanbul ignore next */
    path = decodeURIComponent(path);
  }

  let segments = splitBySlash(path).filter(Boolean);
  let state = S.create(config);

  if (segments.length === 0) {
    state = S.advanceToFirst(state);
    return state;
  }

  segments.forEach(segment => {
    const m = parsePathSegment(segment);
    if (m == null) {
      return;
    }
    let {actionId, commandName, args} = m;
    state = S.advanceTo(state, actionId);
    if (args || commandName) {
      commandName = commandName || 'default';
      state = S.applyCommandAtCurrentPosition(state, commandName, splitByComma(args));
    }
  });

  return state;
}

/**
 * Serialize ``graph`` object to string suitable to be used as path.
 */
export function toPath(state: S.State): string {
  invariant(state.position != null, 'Invalid state');
  const breadcrumb = state.position.trace.concat(state.position);
  let path = '/' + joinWithSlash(breadcrumb.map(positionToPathSegment));
  // We need a workaround as FF doesn't behave consistently.
  return path;
}

const PARSE_SEGMENT = /^([a-zA-Z\-\_]+)(?:\.([a-zA-Z_]+))?(?:\[([^\]]+)\])?$/;

function parsePathSegment(segment) {
  let m = PARSE_SEGMENT.exec(segment);
  if (!m) {
    return null;
  }
  let [_, actionId, commandName, args] = m;
  return {actionId, commandName, args};
}

function positionToPathSegment(pos: S.Position): string {
  let segment = pos.instruction.action.id;
  if (hasCommand(pos)) {
    let {commandName, args} = (pos.command: any);
    const action = pos.instruction.action.element;
    const command = S.getActionCommand(pos.instruction.action, commandName);
    invariant(command != null, 'Unable to find command: %s', commandName);
    let commandArgs = args.map((arg, idx) =>
      command.argumentTypes[idx].stringify(action.props, arg));
    if (commandName === 'default') {
      segment = segment + `[${joinWithComma(commandArgs)}]`;
    } else {
      segment = segment + `.${commandName}[${joinWithComma(commandArgs)}]`;
    }
  }
  return segment;
}

function hasCommand(pos: S.Position): boolean {
  const command = pos.command;
  return command != null &&
    (command.args.length === 0 ||
      (command.args.length > 0 && command.args.some(arg => arg)));
}
