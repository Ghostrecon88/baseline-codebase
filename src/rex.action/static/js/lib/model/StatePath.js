/**
 * @copyright 2016-present, Prometheus Research, LLC
 * @flow
 */

import invariant from 'invariant';
import createLogger from 'debug';

import {splitBySlash, splitByComma, joinWithSlash, joinWithComma} from '../StringUtils';
import * as Environment from '../Environment';
import type {StateConfig, State, Position} from './types';
import * as S from './State';
import * as P from './Position';

const log = createLogger('rex-action:model:StatePath');

/**
 * Deserialize state from `path`.
 */
export function fromPath(path: string, config: StateConfig): State {
  // We need a workaround as FF doesn't behave consistently.
  log('fromPath', path);
  if (Environment.isFirefox()) {
    /* istanbul ignore next */
    path = decodeURIComponent(path);
  }

  let segments = splitBySlash(path).filter(Boolean);
  let state = S.create(config);

  if (segments.length === 0) {
    state = S.advanceToFirst(state);
    return state;
  }

  for (const segment of segments) {
    const m = parsePathSegment(segment);
    if (m == null) {
      break;
    }
    let {actionName, commandName, args} = m;
    state = S.advanceTo(state, actionName, null, {allowInvalidPositions: true});
    if (args || commandName) {
      commandName = commandName || 'default';
      state = S.applyCommandAtCurrentPosition(state, commandName, splitByComma(args));
    }
  }

  return state;
}

/**
 * Serialize ``graph`` object to string suitable to be used as path.
 */
export function toPath(state: State): string {
  invariant(state.position != null, 'Invalid state');
  const breadcrumb = S.breadcrumb(state);
  let path = '/' + joinWithSlash(breadcrumb.map(positionToPathSegment));
  // We need a workaround as FF doesn't behave consistently.
  return path;
}

const PARSE_SEGMENT = /^([a-zA-Z\-\_]+)(?:\.([a-zA-Z_]+))?(?:\[([^\]]+)\])?$/;

function parsePathSegment(segment) {
  let m = PARSE_SEGMENT.exec(segment);
  if (!m) {
    return null;
  }
  let [_, actionName, commandName, args] = m;
  return {actionName, commandName, args};
}

function positionToPathSegment(pos: Position): string {
  let segment = pos.instruction.action.name;
  if (hasCommand(pos)) {
    let {commandName, args} = (pos.command: any);
    const action = pos.instruction.action.element;
    const command = P.getCommandForPosition(pos, commandName);
    invariant(command != null, 'Unable to find command: %s', commandName);
    let commandArgs = args.map((arg, idx) =>
      command.argumentTypes[idx].stringify(action.props, arg));
    if (commandName === 'default') {
      segment = segment + `[${joinWithComma(commandArgs)}]`;
    } else {
      segment = segment + `.${commandName}[${joinWithComma(commandArgs)}]`;
    }
  }
  return segment;
}

function hasCommand(pos: Position): boolean {
  const command = pos.command;
  return command != null &&
    (command.args.length === 0 ||
      (command.args.length > 0 && command.args.some(arg => arg)));
}
