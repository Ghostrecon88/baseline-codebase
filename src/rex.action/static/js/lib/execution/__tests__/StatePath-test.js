/**
 * @flow
 */

import * as React from 'react';
import * as SP from '../StatePath';
import * as S from '../State';
import * as C from '../Command';
import * as T from '../../Typing';

describe('StatePath.fromPath()/toPath()', function() {
  const contextTypes = {
    input: new T.RecordType({}),
    output: new T.RecordType({}),
  };
  const setValueCommand = new C.Command(
    function(props, context, value) {
      return {...context, value};
    },
    'default',
    [C.Types.Value()],
  );
  function createAction({id, commands}) {
    return {
      id,
      domain: {},
      contextTypes,
      element: <div />,
      commands,
    };
  }
  const instruction: S.Instruction[] = [
    {
      type: 'execute',
      action: createAction({id: 'one', commands: {}}),
      then: [],
    },
    {
      type: 'execute',
      action: createAction({id: 'two', commands: {}}),
      then: [
        {
          type: 'execute',
          action: createAction({id: 'three', commands: {}}),
          then: [],
        },
      ],
    },
    {
      type: 'execute',
      action: createAction({
        id: 'with-default-command',
        commands: {
          default: setValueCommand,
        },
      }),
      then: [],
    },
    {
      type: 'execute',
      action: createAction({
        id: 'with-named-command',
        commands: {
          named: setValueCommand,
        },
      }),
      then: [],
    },
  ];
  const config = {instruction, context: {}};

  test('', function() {
    const state = SP.fromPath('', config);
    expect(state.position).toMatchSnapshot();
    const path = SP.toPath(state);
    expect(path).toBe('/one');
  });

  test('/', function() {
    const state = SP.fromPath('/', config);
    expect(state.position).toMatchSnapshot();
    const path = SP.toPath(state);
    expect(path).toBe('/one');
  });

  test('/one', function() {
    const state = SP.fromPath('/one', config);
    expect(state.position).toMatchSnapshot();
    const path = SP.toPath(state);
    expect(path).toBe('/one');
  });

  test('/two', function() {
    const state = SP.fromPath('/two', config);
    expect(state.position).toMatchSnapshot();
    const path = SP.toPath(state);
    expect(path).toBe('/two');
  });

  test('/two/three', function() {
    const state = SP.fromPath('/two/three', config);
    expect(state.position).toMatchSnapshot();
    const path = SP.toPath(state);
    expect(path).toBe('/two/three');
  });

  test('/with-default-command', function() {
    const state = SP.fromPath('/with-default-command', config);
    expect(state.position).toMatchSnapshot();
    const path = SP.toPath(state);
    expect(path).toBe('/with-default-command');
  });

  test('/with-default-command[42]', function() {
    const state = SP.fromPath('/with-default-command[42]', config);
    expect(state.position).toMatchSnapshot();
    const path = SP.toPath(state);
    expect(path).toBe('/with-default-command[42]');
  });

  test('/with-named-command', function() {
    const state = SP.fromPath('/with-named-command', config);
    expect(state.position).toMatchSnapshot();
    const path = SP.toPath(state);
    expect(path).toBe('/with-named-command');
  });

  test('/with-named-command.named[32]', function() {
    const state = SP.fromPath('/with-named-command.named[32]', config);
    expect(state.position).toMatchSnapshot();
    const path = SP.toPath(state);
    expect(path).toBe('/with-named-command.named[32]');
  });
});
