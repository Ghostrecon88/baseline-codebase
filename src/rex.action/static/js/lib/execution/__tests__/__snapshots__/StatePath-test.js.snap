// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatePath.fromPath()/toPath()  1`] = `
Object {
  "context": Object {},
  "instruction": Object {
    "action": Object {
      "commands": Object {},
      "contextTypes": Object {
        "input": RecordType {
          "open": true,
          "rows": Object {},
        },
        "output": RecordType {
          "open": true,
          "rows": Object {},
        },
      },
      "domain": Object {},
      "element": <div />,
      "id": "one",
    },
    "then": Array [],
    "type": "execute",
  },
  "stack": Array [],
  "state": Object {},
  "trace": Array [],
}
`;

exports[`StatePath.fromPath()/toPath() / 1`] = `
Object {
  "context": Object {},
  "instruction": Object {
    "action": Object {
      "commands": Object {},
      "contextTypes": Object {
        "input": RecordType {
          "open": true,
          "rows": Object {},
        },
        "output": RecordType {
          "open": true,
          "rows": Object {},
        },
      },
      "domain": Object {},
      "element": <div />,
      "id": "one",
    },
    "then": Array [],
    "type": "execute",
  },
  "stack": Array [],
  "state": Object {},
  "trace": Array [],
}
`;

exports[`StatePath.fromPath()/toPath() /one 1`] = `
Object {
  "context": Object {},
  "instruction": Object {
    "action": Object {
      "commands": Object {},
      "contextTypes": Object {
        "input": RecordType {
          "open": true,
          "rows": Object {},
        },
        "output": RecordType {
          "open": true,
          "rows": Object {},
        },
      },
      "domain": Object {},
      "element": <div />,
      "id": "one",
    },
    "then": Array [],
    "type": "execute",
  },
  "stack": Array [],
  "state": Object {},
  "trace": Array [],
}
`;

exports[`StatePath.fromPath()/toPath() /two 1`] = `
Object {
  "context": Object {},
  "instruction": Object {
    "action": Object {
      "commands": Object {},
      "contextTypes": Object {
        "input": RecordType {
          "open": true,
          "rows": Object {},
        },
        "output": RecordType {
          "open": true,
          "rows": Object {},
        },
      },
      "domain": Object {},
      "element": <div />,
      "id": "two",
    },
    "then": Array [
      Object {
        "action": Object {
          "commands": Object {},
          "contextTypes": Object {
            "input": RecordType {
              "open": true,
              "rows": Object {},
            },
            "output": RecordType {
              "open": true,
              "rows": Object {},
            },
          },
          "domain": Object {},
          "element": <div />,
          "id": "three",
        },
        "then": Array [],
        "type": "execute",
      },
    ],
    "type": "execute",
  },
  "stack": Array [],
  "state": Object {},
  "trace": Array [],
}
`;

exports[`StatePath.fromPath()/toPath() /two/three 1`] = `
Object {
  "context": Object {},
  "instruction": Object {
    "action": Object {
      "commands": Object {},
      "contextTypes": Object {
        "input": RecordType {
          "open": true,
          "rows": Object {},
        },
        "output": RecordType {
          "open": true,
          "rows": Object {},
        },
      },
      "domain": Object {},
      "element": <div />,
      "id": "three",
    },
    "then": Array [],
    "type": "execute",
  },
  "stack": Array [],
  "state": Object {},
  "trace": Array [
    Object {
      "context": Object {},
      "instruction": Object {
        "action": Object {
          "commands": Object {},
          "contextTypes": Object {
            "input": RecordType {
              "open": true,
              "rows": Object {},
            },
            "output": RecordType {
              "open": true,
              "rows": Object {},
            },
          },
          "domain": Object {},
          "element": <div />,
          "id": "two",
        },
        "then": Array [
          Object {
            "action": Object {
              "commands": Object {},
              "contextTypes": Object {
                "input": RecordType {
                  "open": true,
                  "rows": Object {},
                },
                "output": RecordType {
                  "open": true,
                  "rows": Object {},
                },
              },
              "domain": Object {},
              "element": <div />,
              "id": "three",
            },
            "then": Array [],
            "type": "execute",
          },
        ],
        "type": "execute",
      },
      "stack": Array [],
      "state": Object {},
      "trace": Array [],
    },
  ],
}
`;

exports[`StatePath.fromPath()/toPath() /with-default-command 1`] = `
Object {
  "context": Object {},
  "instruction": Object {
    "action": Object {
      "commands": Object {
        "default": Command {
          "_execute": [Function],
          "argumentTypes": Array [
            ValueType {},
          ],
          "name": "default",
        },
      },
      "contextTypes": Object {
        "input": RecordType {
          "open": true,
          "rows": Object {},
        },
        "output": RecordType {
          "open": true,
          "rows": Object {},
        },
      },
      "domain": Object {},
      "element": <div />,
      "id": "with-default-command",
    },
    "then": Array [],
    "type": "execute",
  },
  "stack": Array [],
  "state": Object {},
  "trace": Array [],
}
`;

exports[`StatePath.fromPath()/toPath() /with-default-command[42] 1`] = `
Object {
  "command": Object {
    "args": Array [
      "42",
    ],
    "commandName": "default",
  },
  "context": Object {
    "value": "42",
  },
  "instruction": Object {
    "action": Object {
      "commands": Object {
        "default": Command {
          "_execute": [Function],
          "argumentTypes": Array [
            ValueType {},
          ],
          "name": "default",
        },
      },
      "contextTypes": Object {
        "input": RecordType {
          "open": true,
          "rows": Object {},
        },
        "output": RecordType {
          "open": true,
          "rows": Object {},
        },
      },
      "domain": Object {},
      "element": <div />,
      "id": "with-default-command",
    },
    "then": Array [],
    "type": "execute",
  },
  "stack": Array [],
  "state": Object {},
  "trace": Array [],
}
`;

exports[`StatePath.fromPath()/toPath() /with-named-command 1`] = `
Object {
  "context": Object {},
  "instruction": Object {
    "action": Object {
      "commands": Object {
        "named": Command {
          "_execute": [Function],
          "argumentTypes": Array [
            ValueType {},
          ],
          "name": "default",
        },
      },
      "contextTypes": Object {
        "input": RecordType {
          "open": true,
          "rows": Object {},
        },
        "output": RecordType {
          "open": true,
          "rows": Object {},
        },
      },
      "domain": Object {},
      "element": <div />,
      "id": "with-named-command",
    },
    "then": Array [],
    "type": "execute",
  },
  "stack": Array [],
  "state": Object {},
  "trace": Array [],
}
`;

exports[`StatePath.fromPath()/toPath() /with-named-command.named[32] 1`] = `
Object {
  "command": Object {
    "args": Array [
      "32",
    ],
    "commandName": "named",
  },
  "context": Object {
    "value": "32",
  },
  "instruction": Object {
    "action": Object {
      "commands": Object {
        "named": Command {
          "_execute": [Function],
          "argumentTypes": Array [
            ValueType {},
          ],
          "name": "default",
        },
      },
      "contextTypes": Object {
        "input": RecordType {
          "open": true,
          "rows": Object {},
        },
        "output": RecordType {
          "open": true,
          "rows": Object {},
        },
      },
      "domain": Object {},
      "element": <div />,
      "id": "with-named-command",
    },
    "then": Array [],
    "type": "execute",
  },
  "stack": Array [],
  "state": Object {},
  "trace": Array [],
}
`;
